src/Microsoft.VisualStudio.LanguageServices.Razor/ProjectSystem/RazorProjectHost.cs

```diff
@@ -0,0 +1,182 @@
+            {
+                throw new ArgumentNullException(nameof(commonServices));
+            }
+
+            if (workspace == null)
+            {
+                throw new ArgumentNullException(nameof(workspace));
+            }
+
+            _commonServices = commonServices;
+            _workspace = workspace;
+
+            _lock = new AsyncSemaphore(initialCount: 1);
+        }
+
+        // Internal for testing
+        internal RazorProjectHost(
+            IUnconfiguredProjectCommonServices commonServices,
+             Workspace workspace,
+             ProjectSnapshotManagerBase projectManager)
+            : base(commonServices.ThreadingService.JoinableTaskContext)
+        {
+            if (commonServices == null)
+            {
+                throw new ArgumentNullException(nameof(commonServices));
+            }
+
+            if (workspace == null)
+            {
+                throw new ArgumentNullException(nameof(workspace));
+            }
+
+            if (projectManager == null)
+            {
+                throw new ArgumentNullException(nameof(projectManager));
+            }
+
+            _commonServices = commonServices;
+            _workspace = workspace;
+            _projectManager = projectManager;
+
+            _lock = new AsyncSemaphore(initialCount: 1);
+        }
+
+
+        [AppliesTo("DotNetCoreRazor")]
+        [ProjectAutoLoad(ProjectLoadCheckpoint.ProjectFactoryCompleted)]
+        public Task LoadAsync()
+        {
+            return InitializeAsync();
+        }
+
+        // Should only be called from the UI thread.
+        private ProjectSnapshotManagerBase GetProjectManager()
+        {
+            _commonServices.ThreadingService.VerifyOnUIThread();
+
+            if (_projectManager == null)
+            {
+                _projectManager = (ProjectSnapshotManagerBase)_workspace.Services.GetLanguageServices(RazorLanguage.Name).GetRequiredService<ProjectSnapshotManager>();
+            }
+
+            return _projectManager;
+        }
+
+        protected override Task InitializeCoreAsync(CancellationToken cancellationToken)
+        {
+            // Don't try to evaluate any properties here since the project is still loading and we require access
+            // to the UI thread to push our updates.
+            //
+            // Just subscribe and handle the notification later.
+            var receiver = new ActionBlock<IProjectVersionedValue<IProjectSubscriptionUpdate>>(OnProjectChanged);
+            _subscription = _commonServices.ActiveConfiguredProjectSubscription.JointRuleSource.SourceBlock.LinkTo(
+                receiver,
+                initialDataAsNew: true,
+                suppressVersionOnlyUpdates: true,
+                ruleNames: new string[] { RazorGeneral.SchemaName, });
+
+            return Task.CompletedTask;
+        }
+
+        protected override async Task DisposeCoreAsync(bool initialized)
+        {
+            _subscription.Dispose();
```

Comment:
I assume InitializeCoreAsycn is guaranteed to have run yes? I only ask because it's not our code.

Reply:
I can make this more defensive.

Reply:
Yeah, there's a pattern for this.. I'm supposed to look at `initialized` - lol

Reply:
Glad I asked 😄